File: address-geocode-us-dot-net/REST/AGUSResponse.cs
using System.Runtime.Serialization;

namespace address_geocode_us_dot_net.REST
{

    /// <summary>
    /// Response from GetBestMatch_V4 API
    /// </summary>
    [DataContract]

    public class GetBestMatchV4Response
    {
        public string Level { get; set; }
        public string LevelDescription { get; set; }
        public string Latitude { get; set; }
        public string Longitude { get; set; }
        public string Zip { get; set; }
        public InformationComponent[] InformationComponents { get; set; }
        public Error Error { get; set; }
        public override string ToString()
        {
            string infoComponentsString = InformationComponents != null
                ? string.Join(", ", InformationComponents.Select(ic => ic.ToString()))
                : "null";

            return $"Level: {Level}\n" +
                   $"LevelDescription: {LevelDescription}\n" +
                   $"Latitude: {Latitude}\n" +
                   $"Longitude: {Longitude}\n" +
                   $"Zip: {Zip}\n" +
                   $"InformationComponents: [{infoComponentsString}]\n" +
                   $"Error: {(Error != null ? Error.ToString() : "null")}";
        }
    }

    /// <summary>
    /// Information Component for GetBestMatch_V4
    /// </summary>
    [DataContract]
    public class InformationComponent
    {
        public string Name { get; set; }
        public string Value { get; set; }
        public override string ToString()
        {
            return $"Name: {Name}, Value: {Value}";
        }
    }


    /// <summary>
    /// Response from GetDistanceToWater API, containing the estimated distance to the nearest saltwater
    /// and the coordinates of the closest saltwater location.
    /// </summary>
    [DataContract]
    public class GetDistanceToWaterResponse
    {
        public string DistanceToWater { get; set; }
        public string Latitude { get; set; }
        public string Longitude { get; set; }
        public string WaterLat { get; set; }
        public string WaterLon { get; set; }
        public Error Error { get; set; }

        public override string ToString()
        {
            return $"MilesToWater: {DistanceToWater}\n" +
                   $"Latitude: {Latitude}\n" +
                   $"Longitude: {Longitude}\n" +
                   $"ClosestWaterLatitude: {WaterLat}\n" +
                   $"ClosestWaterLongitude: {WaterLon}\n" +
                   $"Error: {(Error != null ? Error.ToString() : "null")}";
        }
    }


    /// <summary>
    /// Response from GetReverseLocation API, containing the estimated address, city, state, ZIP code,
    /// and county for the given coordinates.
    /// </summary>
    [DataContract]
    public class GetReverseLocationResponse
    {
        public string Address { get; set; }
        public string City { get; set; }
        public string State { get; set; }
        public string Zip { get; set; }
        public string County { get; set; }
        public Error Error { get; set; }
        public override string ToString()
        {
            return $"Address: {Address}\n" +
                   $"City: {City}\n" +
                   $"State: {State}\n" +
                   $"Zip: {Zip}\n" +
                   $"County: {County}\n" +
                   $"Error: {(Error != null ? Error.ToString() : "null")}";
        }
    }
    /// <summary>
    /// Error object for API responses
    /// </summary>
    [DataContract]
    public class Error
    {
        public string Desc { get; set; }
        public string Number { get; set; }
        public string Location { get; set; }
        public override string ToString()
        {
            return $"Desc: {Desc}, TypeCode: {Number}, Location: {Location}";
        }
    }
}

File: address-geocode-us-nodejs/SOAP/agus_response.js
export class GetDistanceToWaterInput {
    constructor(data = {}) {
        this.Latitude = data.Latitude || null;
        this.Longitude = data.Longitude || null;
        this.LicenseKey = data.LicenseKey || null;
        this.IsLive = data.IsLive !== undefined ? data.IsLive : true;
        this.TimeoutSeconds = data.TimeoutSeconds !== undefined ? data.TimeoutSeconds : 15;
    }

    toString() {
        return `GetDistanceToWaterInput: Latitude = ${this.Latitude}, Longitude = ${this.Longitude}, LicenseKey = ${this.LicenseKey}, IsLive = ${this.IsLive}, TimeoutSeconds = ${this.TimeoutSeconds}`;
    }
}

/**
 * Information Component for API responses.
 */
export class InformationComponent {
    constructor(data = {}) {
        this.Name = data.Name || null;
        this.Value = data.Value || null;
    }

    toString() {
        return `Name = ${this.Name}, Value = ${this.Value}`;
    }
}

/**
 * Error object for API responses.
 */
export class Error {
    constructor(data = {}) {
        this.Desc = data.Desc || null;
        this.Number = data.Number || null;
        this.Location = data.Location || null;
    }

    toString() {
        return `Error: Desc = ${this.Desc}, Number = ${this.Number}, Location = ${this.Location}`;
    }
}

/**
 * Response from GetBestMatch_V4 API.
 */
export class GetBestMatchV4Response {
    constructor(data = {}) {
        this.Level = data.Level || null;
        this.LevelDescription = data.LevelDescription || null;
        this.Latitude = data.Latitude || null;
        this.Longitude = data.Longitude || null;
        this.Zip = data.Zip || null;
        this.InformationComponents =data.InformationComponents.InformationComponent;
        this.Error = data.Error ? new Error(data.Error) : null;
    }

    toString() {
        const componentsString = this.InformationComponents.length
            ? this.InformationComponents.map(component => component.toString()).join(', ')
            : 'null';
        return `GetBestMatchV4Response: Level = ${this.Level}, LevelDescription = ${this.LevelDescription}, Latitude = ${this.Latitude}, Longitude = ${this.Longitude}, Zip = ${this.Zip}, InformationComponents = [${componentsString}], Error = ${this.Error ? this.Error.toString() : 'null'}`;
    }
}

/**
 * Response from GetDistanceToWater API, containing the estimated distance to the nearest saltwater
 * and the coordinates of the closest saltwater location.
 */
export class GetDistanceToWaterResponse {
    constructor(data = {}) {
        this.DistanceToWater = data.DistanceToWater || null;
        this.Latitude = data.Latitude || null;
        this.Longitude = data.Longitude || null;
        this.WaterLat = data.WaterLat || null;
        this.WaterLon = data.WaterLon || null;
        this.Error = data.Error ? new Error(data.Error) : null;
    }

    toString() {
        return `GetDistanceToWaterResponse: MilesToWater = ${this.DistanceToWater}, Latitude = ${this.Latitude}, Longitude = ${this.Longitude}, ClosestWaterLatitude = ${this.WaterLat}, ClosestWaterLongitude = ${this.WaterLon}, Error = ${this.Error ? this.Error.toString() : 'null'}`;
    }
}

/**
 * Response from GetReverseLocation API, containing the estimated address, city, state, ZIP code,
 * and county for the given coordinates.
 */
export class GetReverseLocationResponse {
    constructor(data = {}) {
        this.Address = data.Address || null;
        this.City = data.City || null;
        this.State = data.State || null;
        this.Zip = data.Zip || null;
        this.County = data.County || null;
        this.Error = data.Error ? new Error(data.Error) : null;
    }

    toString() {
        return `GetReverseLocationResponse: Address = ${this.Address}, City = ${this.City}, State = ${this.State}, Zip = ${this.Zip}, County = ${this.County}, Error = ${this.Error ? this.Error.toString() : 'null'}`;
    }
}

export default GetDistanceToWaterResponse;